File: Blanks/WordModel.m
//
//  WordModel.m
//  Blanks
//
//  Created by Kai Kunze on 05.08.12.
//  Copyright (c) 2012 Kai Kunze. All rights reserved.
//

#import "WordModel.h"

@implementation WordModel

@synthesize correct,definition,difficulty;

-(id) init{
	self=[super init];
    if (self) {
        srandom(time(NULL));
        NSString *filename =@"average";
        difficulty = [[NSUserDefaults standardUserDefaults] objectForKey: @"Difficulty"];
        if(!difficulty){
            difficulty =@"average";
            [[NSUserDefaults standardUserDefaults] setObject: difficulty forKey: @"Difficulty"];
        }else{
            filename = difficulty;
        }
        //XXX TODO fix filename stuff
        filename = @"average";
        NSString *path = [[NSBundle mainBundle]  pathForResource:filename ofType:@"plist"];
        wordsArray = [[NSMutableArray alloc] initWithContentsOfFile:path];
        
        int randWordID = random() % [wordsArray count];
        NSMutableDictionary *element = [wordsArray objectAtIndex:randWordID];
        self.correct=[element objectForKey:(id)@"word"];
        self.definition=[element objectForKey:(id)@"definition"];
        falseWordArray=[element objectForKey:(id)@"false"];
        return self;

    }
    return self;
}

-(BOOL) didListChange{
	NSString *newDifficulty = [[NSUserDefaults standardUserDefaults] objectForKey: @"Difficulty"];
	return ![difficulty isEqualToString:newDifficulty];
}

///XXX TODO  XXX fix duplicate code
-(void) checkOptions{
	NSString *newDifficulty = [[NSUserDefaults standardUserDefaults] objectForKey: @"Difficulty"];
	if( ![difficulty isEqualToString:newDifficulty]){
		difficulty = newDifficulty;
		NSString *path = [[NSBundle mainBundle]  pathForResource:difficulty ofType:@"plist"];
		wordsArray = [[NSMutableArray alloc] initWithContentsOfFile:path];
		srandom(time(NULL));
        [self selectNextWord];
	}
}
-(void) selectNextWord{
    int randWordID = random() % [wordsArray count];
	NSMutableDictionary *element = [wordsArray objectAtIndex:randWordID];
	self.correct=[element objectForKey:(id)@"word"];
	self.definition=[element objectForKey:(id)@"definition"];
	falseWordArray=[element objectForKey:(id)@"false"];
	//NSLog(@"correct %@, definition %@", correct, definition);
	
}


-(NSArray*) getWords{
	NSMutableArray *words= [[NSMutableArray alloc] initWithCapacity:4];
	int first = random() % [falseWordArray count];
	int second = random() % [falseWordArray count];
	while(first == second){
		second = random() % [falseWordArray count];
	}
	int third = random() % [falseWordArray count];
	while((third == first) ||(third == second)){
		third = random() % [falseWordArray count];
	}
	
	[words insertObject:[falseWordArray objectAtIndex:first] atIndex:0];
	[words insertObject:[falseWordArray objectAtIndex:second] atIndex:1];
	[words insertObject:[falseWordArray objectAtIndex:third] atIndex:2];
	[words insertObject:correct atIndex:3];
    
	int firstObject = 0;
	for (int i = 0; i<[words count];i++) {
		NSUInteger randomIndex = random() % [words count];
		[words exchangeObjectAtIndex:firstObject withObjectAtIndex:randomIndex];
		firstObject +=1;
	}
	
	return words;
}



@end

File: Blanks/WordModel.swift
//
//  WordModel.swift
//  Blanks
//
//  Created by Kai Kunze on 05.08.12.
//  Copyright (c) 2012 Kai Kunze. All rights reserved.
//

import Foundation

class WordModel: NSObject {
    private(set) var correct: String = ""
    private(set) var definition: String = ""
    private var difficulty: String = "average"

    private var wordsArray: [[String: Any]] = []
    private var falseWordArray: [String] = []

    override init() {
        super.init()

        // Load difficulty from UserDefaults
        if let savedDifficulty = UserDefaults.standard.string(forKey: .difficulty) {
            difficulty = savedDifficulty
        } else {
            difficulty = DifficultyLevel.average.rawValue
            UserDefaults.standard.set(difficulty, forKey: .difficulty)
        }

        // XXX TODO fix filename stuff
        difficulty = DifficultyLevel.average.rawValue

        // Load words for current difficulty
        loadWordList(for: difficulty)

        // Select initial random word
        selectNextWord()
    }

    func checkOptions() {
        if let newDifficulty = UserDefaults.standard.string(forKey: .difficulty),
           difficulty != newDifficulty {
            difficulty = newDifficulty
            loadWordList(for: difficulty)
            selectNextWord()
        }
    }

    func selectNextWord() {
        guard !wordsArray.isEmpty else { return }

        let randWordID = Int.random(in: 0..<wordsArray.count)
        let element = wordsArray[randWordID]
        self.correct = element["word"] as? String ?? ""
        self.definition = element["definition"] as? String ?? ""
        self.falseWordArray = element["false"] as? [String] ?? []
    }

    func getWords() -> [String] {
        var words: [String] = []

        guard falseWordArray.count >= 3 else {
            return [correct]
        }

        // Select 3 unique false words
        var indices: Set<Int> = []
        while indices.count < 3 {
            let randomIndex = Int.random(in: 0..<falseWordArray.count)
            indices.insert(randomIndex)
        }

        let indicesArray = Array(indices)
        words.append(falseWordArray[indicesArray[0]])
        words.append(falseWordArray[indicesArray[1]])
        words.append(falseWordArray[indicesArray[2]])
        words.append(correct)

        // Shuffle the array
        words.shuffle()

        return words
    }

    // MARK: - Private Helpers

    private func loadWordList(for difficulty: String) {
        if let path = Bundle.main.path(forResource: difficulty, ofType: "plist"),
           let array = NSArray(contentsOfFile: path) as? [[String: Any]] {
            wordsArray = array
        }
    }
}
